src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -154076,5 +154122,1342 @@ class C { }
+                // (18,44): warning CS8619: Nullability of reference types in value of type 'C2<string>' doesn't match target type 'C2<string?>'.
+                //         c = new() { F1 = GetC2("c1"), F2 = GetC2("c2")};
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, @"GetC2(""c2"")").WithArguments("C2<string>", "C2<string?>").WithLocation(18, 44),
+                // (20,59): warning CS8619: Nullability of reference types in value of type 'C2<string>' doesn't match target type 'C2<string?>'.
+                //         var d = new C1<string?>?(new C1<string?>() { F1 = new C2<string>("d1"), F2 = new C2<string>("d2")});
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, @"new C2<string>(""d1"")").WithArguments("C2<string>", "C2<string?>").WithLocation(20, 59),
+                // (20,86): warning CS8619: Nullability of reference types in value of type 'C2<string>' doesn't match target type 'C2<string?>'.
+                //         var d = new C1<string?>?(new C1<string?>() { F1 = new C2<string>("d1"), F2 = new C2<string>("d2")});
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, @"new C2<string>(""d2"")").WithArguments("C2<string>", "C2<string?>").WithLocation(20, 86),
+                // (22,57): warning CS8619: Nullability of reference types in value of type 'C2<string?>' doesn't match target type 'C2<string>'.
+                //         var e = new C1<string>?(new C1<string>() { F1 = GetC2(s), F2 = GetC2("e")});
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "GetC2(s)").WithArguments("C2<string?>", "C2<string>").WithLocation(22, 57),
+                // (22,72): warning CS8619: Nullability of reference types in value of type 'C2<string>' doesn't match target type 'C2<string?>'.
+                //         var e = new C1<string>?(new C1<string>() { F1 = GetC2(s), F2 = GetC2("e")});
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, @"GetC2(""e"")").WithArguments("C2<string>", "C2<string?>").WithLocation(22, 72),
+                // (24,9): warning CS8602: Dereference of a possibly null reference.
+                //         e.Value.F2.F3.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "e.Value.F2.F3").WithLocation(24, 9)
+                );
+        }
+
+        [Fact]
+        public void ObjectCreationExpression_TargetTyped_10()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    static void M()
+    {
+        var b = GetC(""b"");
+        b = new() { null, ""b""};
+
+        var c = new C<string>?(new C<string>() { null, ""c""});
+        C<string>? d = new() { null, ""d""};
+    }
+
+    static C<T>? GetC<T>(T t) => new C<T>(t);
+}
+
+struct C<T> : System.Collections.IEnumerable
+{
+    public C(){}
+    public C(T t){}
+    public void Add(T t){}
```

Comment:
Consider a variation on this test where `Add` is a generic method constrained by `T`: something like `Add<U>(U u) where U : T`.

Reply:
(And giving the `null`s a type to facilitiate type inference of the `Add` call)

Reply:
Are ```ObjectCreationExpression_TargetTyped_06``` and ```ObjectCreationExpression_TargetTyped_12``` not good enough? I think they involve inference for the ```Add``` call. I guess not exactly what you are asking, but feels sufficient to me. If not, please construct the complete test scenario and what you expect for it and I will be happy to add it as a test.

Reply:
The scenario would be:

```cs
class C
{
    static void M()
    {
        var b = GetC(""b"");
        b = new() { default(string), ""b""};

        var c = new C<string>?(new C<string>() { null, ""c""});
        C<string>? d = new() { default(string), ""d""};
    }

    static C<T>? GetC<T>(T t) => new C<T>(t);
}

struct C<T> : System.Collections.IEnumerable
{
    public C(){}
    public C(T t){}
    public void Add<U>(U u) where U : T {}

    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => throw null!;
}
```

I'd expect to see the same warnings as in this test case: What I'm thinking is interesting about it (and different from _06 and _12) is that the `Add` method needs to be retargeted to the new `C<string>` type, and then have method inference run over it.

Reply:
> I'd expect to see the same warnings as in this test case

Not quite the same, but still warnings:
```
// (7,21): warning CS8631: The type 'string?' cannot be used as type parameter 'U' in the generic type or method 'C<string>.Add<U>(U)'. Nullability of type argument 'string?' doesn't match constraint type 'string'.
//         b = new() { default(string), "b"};
Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "default(string)").WithArguments("C<string>.Add<U>(U)", "string", "U", "string?").WithLocation(7, 21),
// (9,50): error CS0411: The type arguments for method 'C<string>.Add<U>(U)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
//         var c = new C<string>?(new C<string>() { null, "c"});
Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "null").WithArguments("C<string>.Add<U>(U)").WithLocation(9, 50),
// (10,32): warning CS8631: The type 'string?' cannot be used as type parameter 'U' in the generic type or method 'C<string>.Add<U>(U)'. Nullability of type argument 'string?' doesn't match constraint type 'string'.
//         C<string>? d = new() { default(string), "d"};
Diagnostic(ErrorCode.WRN_NullabilityMismatchInTypeParameterConstraint, "default(string)").WithArguments("C<string>.Add<U>(U)", "string", "U", "string?").WithLocation(10, 32)
```

Does this match your expectations?

Reply:
I guess you meant to use ```default(string)``` everywhere where ```null``` was used.

Reply:
Yeah, I wasn't entirely sure whether it was going to be the mismatched type parameter warning or the possible null reference parameter warning. Thanks.

Reply:
> I guess you meant to use `default(string)` everywhere where `null` was used.

Yeah, I missed one in my example.